waspc/data/Generator/templates/react-app/src/auth/pages/OAuthCallback.tsx

```diff
@@ -0,0 +1,82 @@
+{{={= =}=}}
+import { useEffect, useRef, useState } from "react";
+import { Redirect } from 'react-router-dom'
+import { useAuth } from 'wasp/client/auth'
+import { api } from 'wasp/client/api'
+import { initSession } from 'wasp/auth/helpers/user'
+
+export function OAuthCallbackPage() {
+  const { isLoading, error, user } = useOAuthCallbackHandler();
+  
+  if (user) {
+    return <Redirect to="{= onAuthSucceededRedirectTo =}" />;
+  }
+
+  return (
+    <div style={{
+      display: "flex",
+      alignItems: "center",
+      justifyContent: "center",
+      padding: "4rem",
+    }}>
+      {error && <div style={{
+        color: "rgb(239 68 68)"
+      }}>{error}</div>}
+      {isLoading && <div>Please wait a moment while we log you in.</div>}
+    </div>
+  );
+}
+
+function useOAuthCallbackHandler() {
+  const { data: user } = useAuth();
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<string | null>(null);
+
+  async function handleCallback() {
+    try {
+      setIsLoading(true);
+      const query = new URLSearchParams(window.location.search);
+
+      // If we got redirect with an error, display it to the user
+      // and don't continue with the login process.
+      if (query.get('error')) {
+        setError(query.get('error'));
+        return;
+      }
+
+      const code = window.location.hash.slice(1);
+      const response = await exchangeOAuthCodeForToken({ code });
+      const sessionId = response.data.sessionId;
+      if (!sessionId) {
+        setError("Unable to login with the OAuth provider.");
+      }
+      await initSession(sessionId);
+    } catch (e: unknown) {
+      console.error(e);
+      setError("Unable to login with the OAuth provider.");
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  const isFirstRender = useRef(true);
+  useEffect(() => {
+    if (isFirstRender.current) {
+      isFirstRender.current = false;
+      handleCallback();
+    }
+  }, []);
```

Comment:
This is a unique hack that we'll only use here. In React 18 with strict mode, each component renders twice. This makes this page problematic since it will send the OAuth code twice (once successfully, once with an error since the OAuth code can be used only once).